### PR TITLE
Make BatchIteratorTester generic; Move some BatchIterators methods

### PR DESCRIPTION
**Pull request by mfussenegger**
See commits

### CODE REVIEW

File: libs/dex/src/main/java/io/crate/data/SentinelRow.java [hunk 1]
```diff
@@ -39,7 +39,7 @@ public Object get(int index) {
 
         @Override
         public Object[] materialize() {
-            throw new IllegalStateException("Iterator is not on a row");
+            return new Object[0];
         }
```
Comment (BaurzhanSakhariev) [inline libs/dex/src/main/java/io/crate/data/SentinelRow.java:43]: What's the purpose of this change? 

I wonder if not reflecting this change in numColumns/get can lead to sonme issues because of mixed semantics?

**Review by BaurzhanSakhariev — APPROVED** [review]
LGTM, left a question

File: libs/dex/src/main/java/io/crate/data/SentinelRow.java (hunk 1 above)
Comment (mfussenegger) [reply libs/dex/src/main/java/io/crate/data/SentinelRow.java:43]: The BatchIteratorTester now uses .map to go from Row -> Object[] via materialize and it asserts that `currentElement` can be called in some off-row scenarios or after close. 

> I wonder if not reflecting this change in numColumns/get can lead to sonme issues because of mixed semantics?

Should be fine as the SentinelRow is only used if in an off position.
